src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts

```diff
@@ -0,0 +1,210 @@
+    const osmosisAdapter = adapterManager.get(osmosisChainId) as osmosis.ChainAdapter | undefined
+    const cosmosAdapter = adapterManager.get(cosmosChainId) as cosmos.ChainAdapter | undefined
+
+    const { REACT_APP_OSMOSIS_NODE_URL: osmoUrl } = getConfig()
+
+    if (!cosmosAdapter || !osmosisAdapter) throw new Error('Failed to get adapters')
+
+    let sellAddress
+
+    if (sellAssetIsOnOsmosisNetwork) {
+      sellAddress = from
+
+      if (!sellAddress)
+        throw new SwapError('failed to get osmoAddress', {
+          code: SwapErrorType.EXECUTE_TRADE_FAILED,
+        })
+    } else {
+      /** If the sell asset is not on the Osmosis network, we need to bridge the
+       * asset to the Osmosis network first in order to perform a swap on Osmosis DEX.
+       */
+      sellAddress = from
+
+      if (!sellAddress) throw new Error('Failed to get address')
+
+      const transfer = {
+        sender: sellAddress,
+        receiver: receiveAddress,
+        amount: sellAmountBeforeFeesCryptoBaseUnit,
+      }
+
+      const responseAccount = await cosmosAdapter.getAccount(sellAddress)
+      const ibcAccountNumber = parseInt(responseAccount.chainSpecific.accountNumber || '0')
+
+      const sequence = responseAccount.chainSpecific.sequence || '0'
+
+      const getFeeDataInput: Partial<GetFeeDataInput<OsmosisSupportedChainId>> = {}
+      const ibcFromCosmosFeeData = await cosmosAdapter.getFeeData(getFeeDataInput)
+
+      const unsignedTx = await buildPerformIbcTransferUnsignedTx({
+        input: transfer,
+        adapter: cosmosAdapter,
+        blockBaseUrl: osmoUrl,
+        denom: 'uatom',
+        sourceChannel: COSMO_OSMO_CHANNEL,
+        feeAmount: ibcFromCosmosFeeData.fast.txFee,
+        accountNumber,
+        ibcAccountNumber,
+        sequence,
+        gas: ibcFromCosmosFeeData.fast.chainSpecific.gasLimit,
+        feeDenom: 'uatom',
+      })
+
+      return unsignedTx
+    }
+
+    // TODO(gomes): uncomment me when actually implementing multi-hop
+    // const osmoAddress = sellAssetIsOnOsmosisNetwork ? sellAddress : receiveAddress
+    // const cosmosAddress = sellAssetIsOnOsmosisNetwork ? receiveAddress : sellAddress
+
+    /** At the current time, only OSMO<->ATOM swaps are supported, so this is fine.
+     * In the future, as more Osmosis network assets are added, the buy asset should
+     * be used as the fee asset automatically. See the whitelist of supported fee assets here:
+     * https://github.com/osmosis-labs/osmosis/blob/04026675f75ca065fb89f965ab2d33c9840c965a/app/upgrades/v5/whitelist_feetokens.go
+     */
+
+    const getFeeDataInput: Partial<GetFeeDataInput<OsmosisSupportedChainId>> = {}
+    const swapFeeData = await (sellAssetIsOnOsmosisNetwork
+      ? osmosisAdapter.getFeeData(getFeeDataInput)
+      : cosmosAdapter.getFeeData(getFeeDataInput))
+
+    const feeDenom = sellAssetIsOnOsmosisNetwork
+      ? 'uosmo'
+      : atomOnOsmosisAssetId.split('/')[1].replace(/:/g, '/')
```

Comment:
guh how does this work? should this be done via `fromAssetId` with a `[.., ..].join('/')` so its clear whats going on?

Reply:
Yeah had the same comment on the spookettiness of this in the original PR - this should really just be `ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2` without any programmatic-ness

Reply:
Actually with the multi-hop refactor in https://github.com/shapeshift/web/pull/4905, this isn't needed anymore: a `swap-exact-amount-in` (i.e what we're doing here) is *always* going to be on the Osmosis chain, thus network fees are always going to be in uosmo.
The reason for this was weird branching with the whole IBC transfer/swap flow, which is now made a lot cleaner with the notion of steps/hops.